core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalLogWriter.java

```diff
@@ -149,24 +175,154 @@ public void close() throws IOException {
+        mUfs.renameFile(src, dst);
+      } else {
+        LOG.info("No journal entry found in current journal file {}. Deleting it", src);
+        mUfs.deleteFile(src);
+      }
+    }
+    // Search for and scan the latest COMPLETE journal and find out the sequence number of the
+    // last persisted journal entry, in case no entry has been found in the INCOMPLETE journal.
+    if (lastPersistSeq < 0) {
+      // Re-evaluate snapshot because the incomplete journal will be destroyed if
+      // it does not contain any valid entry.
+      snapshot = UfsJournalSnapshot.getSnapshot(mJournal);
+      // journalFiles[journalFiles.size()-1] is the latest journal file. If we have reached this
+      // point, we must have already deleted the previous empty incomplete journal file.
+      // Therefore, journalFiles[journalFiles.size()-1] must be a valid complete journal file
+      // assuming the Ufs was functioning correctly before journal file rotation.
+      // As a result, we don't have to read the journal file to find out the last journal entry
+      // sequence. We can get this information from the file name. <startSeq>-<endSeq>.
+      List<UfsJournalFile> journalFiles = snapshot.getLogs();
+      if (!journalFiles.isEmpty()) {
+        UfsJournalFile journal = journalFiles.get(journalFiles.size() - 1);
+        lastPersistSeq = journal.getEnd() - 1;
+        LOG.info("Found last persisted journal entry with seq {} in {}.",
+            lastPersistSeq, journal.getLocation().toString());
+      }
+    }
+
+    if (!mEntriesToFlush.isEmpty()) {
+      maybeRotateLog();
+      JournalEntry firstEntryToFlush = mEntriesToFlush.peek();
+      if (firstEntryToFlush.getSequenceNumber() > lastPersistSeq + 1) {
+        throw new RuntimeException(ExceptionMessage.JOURNAL_ENTRY_MISSING.getMessageWithUrl(
+            RuntimeConstants.ALLUXIO_DEBUG_DOCS_URL,
+            lastPersistSeq + 1, firstEntryToFlush.getSequenceNumber()));
+      }
+      long retryStartSeq = firstEntryToFlush.getSequenceNumber();
+      long retryEndSeq = retryStartSeq;
+      LOG.info("Retry writing unwritten journal entries from seq {}", retryStartSeq);
+      for (JournalEntry entry : mEntriesToFlush) {
+        if (entry.getSequenceNumber() > lastPersistSeq) {
+          try {
+            entry.toBuilder().build()
+                .writeDelimitedTo(mJournalOutputStream.mOutputStream);
+            retryEndSeq = entry.getSequenceNumber();
+          } catch (IOException e) {
+            mRotateLogForNextWrite = true;
+            UfsJournalFile tempLog = mJournalOutputStream.mCurrentLog;
+            mJournalOutputStream = null;
+            throw new IOException(ExceptionMessage.JOURNAL_WRITE_FAILURE
+                .getMessageWithUrl(RuntimeConstants.ALLUXIO_DEBUG_DOCS_URL,
+                    tempLog, e.getMessage()), e);
+          }
+        }
+      }
+      LOG.info("Finished writing unwritten journal entries from {} to {}.",
```

Comment:
our earlier logging talks about the last persisted seq number, so it might be confusing to log "found last persisted seq number to be 10", then "wrote unwritten journal entries from 5 to 20". It would be clearer to follow with "re-wrote journal entries from 10 to 20".